### PR TITLE
Manual backport: Use sessions in mongo aggregation pipelines (#35680)

### PR DESCRIPTION
**Pull request by lbrdnk**
Manual backport of #35680. Use sessions in mongo aggregation pipelines.

### CODE REVIEW

File: modules/drivers/mongo/src/metabase/driver/mongo/execute.clj [hunk 1]
```diff
@@ -124,24 +125,27 @@
...
+(defn- init-aggregate!
+  [^AggregateIterable aggregate
+   ^java.lang.Long timeout-ms]
+  (doto aggregate
+    (.allowDiskUse true)
+    ;; TODO - consider what the best batch size option is here. Not sure what the default is.
+    (.batchSize 100)
+    (.maxTime timeout-ms TimeUnit/MILLISECONDS)))
+
+(defn- ^:dynamic *aggregate*
```
Comment (metamben) [inline modules/drivers/mongo/src/metabase/driver/mongo/execute.clj:138]: Why do we prefer making this dynamic?

File: modules/drivers/mongo/src/metabase/driver/mongo/execute.clj [hunk 2]
```diff
@@ -124,24 +125,27 @@
...
+  [^MongoDatabase db
+   ^String coll
+   ^ClientSession session
+   stages timeout-ms]
```
Comment (metamben) [inline modules/drivers/mongo/src/metabase/driver/mongo/execute.clj:142]: This is surprising indentation.

File: modules/drivers/mongo/src/metabase/driver/mongo/execute.clj [hunk 3]
```diff
@@ -156,30 +160,57 @@
...
+      (a/go
+        (when (a/<! (qp.context/canceled-chan context))
+          (kill-session! client-database session)))
+      (let [aggregate ^AggregateIterable (*aggregate* client-database
+                                                      collection-name
+                                                      session
+                                                      query
+                                                      (qp.context/timeout context))]
+        (with-open [^MongoCursor cursor (try (.cursor aggregate)
+                                             (catch Throwable e
```
[the remark this answers is not in the document]
Comment (lbrdnk) [reply modules/drivers/mongo/src/metabase/driver/mongo/execute.clj:210]: I've tried to match the behavior of `sql-jdbc.execute/execute-reducible-query`.

[the remark this answers is not in the document]
Comment (metamben) [reply modules/drivers/mongo/src/metabase/driver/mongo/execute.clj:210]: I don't like that behavior, but let's leave it anyway.

**Review by metamben — APPROVED** [review]

**Pull request merged**